### ISSUE TITLE
support AJAX-y sites

### ISSUE THREAD

**Issue by willkg** [mozilla/mozilla-django-oidc#133]
The current middleware and the new one don't support AJAX-y sites.

This issue covers figuring out how web-sites that are predominantly doing AJAX requests to move data around will work with this library.

**Comment by peterbe** [comment]
@johngian You mentioned *"OpenID connect "implicit flow"."* in https://github.com/mozilla/mozilla-django-oidc/issues/94#issuecomment-305550508
Can you elaborate on that?

**Comment by peterbe** [comment]
@gdestuynder You mentioned *"There is also a draft OIDC spec for session support that also works for backends - but it's still in draft and not supported anywhere yet (AFAIK)** in https://github.com/mozilla/mozilla-django-oidc/issues/94#issuecomment-305648877
So does that mean there is no good way to check if a user is still valid without involving a browser and a redirect?

One option would be, if it's even possible, to have a cron job (or something removed from the web request) that asks the provider if the email is still valid. Considering that Mozilla has a mechanism where a Workday edit (e.g. employee no longer employee) pings LDAP which pings Auth0 (becomes a "blocked user"). Then Auth0 should, ideally, have a way to query "/is-user-block?email=...". 

One usage pattern I'm interested in, because I'm not a big fan of cron jobs, is to make a ping like that from my server when the client makes a request with a custom auth token. E.g. `curl -H "auth-token: mylongtokenhere" https://example.com/api/rest/stuff.json`. It could be something like this:

```python

class AuthTokenMiddleware:
    def process_request(self, request):
         ...if request has header "auth-token" and that checks out...

         cache_key = 'still-valid-user:{}'.format(request.user.id)
         if not cache.get(cache_key):
             if not mozilla_django_oidc.auth.check_still_valid_user(request.user.email):
                 auth.logout(request.user)  # kill django session cookie
                 return HttpForbidden("Your auth-token was found but belongs to a no longer active user according to Auth0")
            cache.set(cache_key, True, 60 * 60 * 24)
        ....
```

Is there anything like that possible? If it's not possible with OIDC, does Auth0 have something custom/proprietary us desperates can use?

**Comment by gdestuynder** [comment]
It is possible to call the Auth0 API backend to get the same information when no redirect flow is possible (with an access token - if it was a cron script it would also have to use an access token anyway).
The main difference though is that it gives the RP access to user's data that the user does not explicitly grants access to (ie bypasses OIDC/OAuth) thus we tend to be careful with it. I've already written code that does similar stuff.

Another solution is to do it in JS like: https://github.com/auth0/auth0.js#api however we haven't tested that yet (and ideally would rather do something that does not depend on Auth0, though maybe we can simulate the same thing they do? This could be the 'best' solution)

**Comment by peterbe** [comment]
I quite like that idea. It's a shame it can't be done with OIDC, but it would definitely solve our "enterprise problem" (i.e. that ownership of email addresses cease some times).
I skimmed the docs real quick and it seems you can use your client domain, client ID and client secret, to request a 24h access token. Then you can do a search by email and the response will have blocked=true|false.

Considering that this feels very Auth0 specific, I guess this thought process can continue in Socorro or Tecken or whatever other service that happens to use Auth0 for their OIDC.

**Comment by gdestuynder** [comment]
That's correct, using your client id/secret to request an access token, then use that to call the Auth0 API.
Your client id needs to be authorized though, but that can be done (its a few clicks on our side). Again we don't want to do this for too many RPs but this is definitely possible (and also is 100% auth0 specific in this case)

The auth0.js solution can possibly be done in a non-auth0 way though there is nothing that I know of that does this out of the box yet.

Another thing I just though of is that https://github.com/pingidentity/mod_auth_openidc **seem** to handle XHR requests properly by using some kind of magic. This may be also worth investigating to see if we can replicate their solution.

**Comment by johngian** [comment]
Although I don't have a solid idea about the implementation details, but does it make sense to implement a custom HTTP API endpoint that responds if the user is blocked/inactive (without revealing any further OIDC data) given an identification attribute (eg. user email) as input?

My thinking behind this is the following:
a) Its not a good idea to request access token and consume data that users haven't agreed to share
b) I don't like implementing stuff outside the OIDC spec because this gives us flexibility in case we want to switch providers
c) This is going to facilitate a small subset of our OIDC clients (RPs in OIDC lingo) that need this extra security feature.

Thoughts?

**Comment by peterbe** [comment]
@johngian 
I think this solution should be entirely outside this project. Once we've built and tested something, we can perhaps turn that into a package or something and release it separately. Then *this* project could *link* to that in its documentation or something. 

@gdestuynder 
The reason I'm excited about the idea of using the Auth0 Management API (isn't that what it's called) is because it'll be useful in two scenarios:
1) An AJAX heavy server that almost always responds with JSON and almost never HTML. 
2) Sites that ad-hoc generates API auth tokens to users. I.e. Socorro. Users then use these API auth tokens in their curl script. Ideally, if they generate an API auth token, write a script that uses it, leaves the company, then the script should stop working tomorrow.

**Comment by gdestuynder** [comment]
For 1. yep - it'd like a non-API solution but the API solution works (and yes its "Auth0 Management API" indeed)

for 2. we have an issue similar to this at https://github.com/mozilla/iam-project-backlog/issues/110, i.e. to be able to issue API tokens (really OAuth access tokens) that are tied to your Auth0 user in a generic-ish way (like, go on your Mozilla SSO dashboard and issue token from your user for RP x with scope Z where the scopes can be defined by the RP - Auth0 has something *close* to that integrated but which requires manual setup and of course relies on the Auth0 API as well)

Basically both points are hot topics we need to solve for, and 1. is one we can solve with the Auth0 API until we find something better

**Comment by peterbe** [comment]
One thing, before I forget, is that we'd need to extend the Hub form to ask "Do you need to use the Auth Management API?"?

@gdestuynder Can you do me a favor? I have a client for symbols.dev.mozaws.net. The client id is hJDCFor3jKW7D7DCt6z4WsaiBOSiIrnb 
Can you upgrade that account so I can use the Auth0 Management API? Then I can write a first draft of this in my Django app. And we can have something concrete to pass around as an example and it might help further the discussion.

**Comment by andrewkrug** [comment]
We should just standardize these scopes so we're being consistent with usage of the management API.

**Comment by gdestuynder** [comment]
For Ajax-y sites I wrote a little POC that makes handling of session easier when the upstream code doesn't do error checking or anything like that:

**Comment by peterbe** [comment]
Because I don't want to forget it; here's an example implementation that does the access token dance and some API ops all written in Python.
https://github.com/mozilla-iam/cis_functions/blob/master/functions/idvtoauth0/authzero.py
Thanks @gdestuynder

**Comment by peterbe** [comment]
So there is now a *real* implementation of this in the wild. [Tecken does it](https://github.com/mozilla-services/tecken/blob/master/tecken/useradmin/middleware.py). It's a middleware that talks to the Auth0 Management API and finds the user and if the user is found **and** is `blocked==true` then it makes the user account inactive, calls `auth.logout(request)`, and immediately returns `403 Forbidden`. It's throttled with the regular `django.core.cache.cache`. It's been looked at by @gdestuynder but no formal security sign-off. 

So, let's put a version of this into `mozilla-django-oidc`. Tecken is probably not the only site that wants to do server-to-server checking to see if someone is still a valid user. 

Points to think about or discuss:

1. This implementation requires that the Auth0 account has been extended with the additional config to allow the client ID to use the Auth0 Management API. (This was done offline by chatting on IRC with kang). This needs to be made much more clear in terms of documentation and possibly push "upstream" so that people can ask for this via The Hub. 

2. Having this extra config on your Auth0 account makes the client ID + secret a lot more powerful so only RPs who are really careful and vetted can/should do this. What needs to be done formally to back this up? We could simply *state* in the documentation "This only works if you have [Auth0 Management API config thing] enabled. Talk to your Auth0 account administrator."

3. Is this much better than doing refresh token redirect loops? If you can use this new trick, what's the point of sending the user on a redirect loop ever X hours? 

4. What should it be called? "auth0specific"? "contrib"? "proprietary?"

5. Is @gdestuynder 's team interested in building a wrapper web service around this to "tighten it down"?

6. If the user signs in once and gets a 1 year session cookie, we lose the ability to pick up refreshed user profile details (avatar, full name etc.) We *could* use the Auth0 Management API to do that. At least make it configurable or something. Is this interesting or a waste of attention?

7. Perhaps it should just log the user out (and return a `403 Forbidden`) and NOT change the user to `is_active=False`. What do you think?

8. Also, I made a [Django management command](https://github.com/mozilla-services/tecken/blob/master/tecken/useradmin/management/commands/is-blocked-in-auth0.py) which was useful during development. Is there interest in including this?

All, in all. Take a look at the implementation and say what you think about having it inside `mozilla-django-oidc`.

**Comment by gdestuynder** [comment]
In general, I would still strongly encourage to use the user-agent redirect loop WITH setting the RP/app cookie expiration to the id token's expiration - that way you never have a situation where a session is incorrectly up (that's a rather bad security risk).
The API mainly allows apps that need to know about users asynchronously, ahead of time, etc. - in fact both can be used at the same time as well.

1. We do have some internal doc on this at least (i.e. "auth0 setup side") at https://mana.mozilla.org/wiki/display/SECURITY/Auth0+Setup#Auth0Setup-TheAuth0managementAPIandAuth0clients though this could be improved of course

3. It depends, the advantage of the user-agent redirect loop is that it requires no additional privileges and there is a clear indication on the auth0 side of what happened (unlike with the API, at the moment at least)

5. Somewhat yes, i.e. we do have a backlog item to do the same thing through our own API instead - others have the same interest and are trying something with a project called Hydra (via @mozbhearsum)

6. We attempt to find any RP that does not follow the rules (ie that uses long sessions but never refresh the user profile in some fashion) - auth0 API is definitely a way that can work, until 5. is solved -albeit it does not give a strong indicator that the user profile is refreshed (and thus, we pick the RP up as "not following the rules" even thus it may do that)

All in all I think this should be optional (due to using a proprietary API + session refresh notification issues) with a clear indication of what it does vs the user-agent redirect - this allows users of the lib to take the best decision for their app.

**Comment by edmorley** [comment]
In case it helps anyone looking at this issue in the future - Treeherder (which is a ~SPA with a django-rest-framework API backend) has just switched to Auth0 and uses the following hybrid approach (for reasoning [see here](https://bugzilla.mozilla.org/show_bug.cgi?id=1395356#c8)):
* the login redirect/callback is handled entirely by the SPA (using a public Auth0 client) - with the Auth0 user session being stored in localstorage / state synced between tabs / session renewal handled silently etc
* the callback page (and later renewals) also sends the Auth0 `access_token` and `id_token` to a django-rest-framework endpoint which calls `authenticate()` on a custom django auth backend
* this auth backend verifies the `id_token`, converts the `subject` and `email` into an appropriate Django username and then handles the rest of the Django login as one might expect, except the session expiry time is overridden to match that from the `access_token` / `id_token`

There are a few nice properties to this:
* the Auth0 access_token can be used client-side to regularly fetch short-lived Taskcluster tokens (required for interacting with third-party APIs), which are intentionally kept client-side only for security
* the custom django auth backend uses the same public Auth0 client for `id_token` verification, so doesn't need an Auth0 client secret - meaning that auth works out of the box in development (without the need for mocking or extra credentials to which contributors don't have access)
* the client side and django sessions now stay in sync (with the previous implementation one would expire before the other, leading to suboptimal UX where the user was told to log out and in again if they needed to perform client-side taskcluster actions)
* so long as the user visits Treeherder once every 24 hours, the silent renewal will keep both the Auth0 and django sessions alive indefinitely (or at least for several weeks I think? not sure if IAM have set a limit on silent renewals?)

For our implementation, see:
mozilla/treeherder#3144

There's probably a fair bit of overlap between our custom django auth backend and parts of this package - perhaps worth seeing if it's possible to reduce the duplication in the future? Though I imagine our use-case may be too specific to be of use to more straight-forwards SPA+django-rest-framework apps (who don't need to manage the Auth0 session client-side at all), so would understand if it's not worth the effort.

**Comment by gdestuynder** [comment]
SPAs with implicit flow are regardless of this less safe and not recommended. An XSS can extract data from localStorage and effectively get access to the id token for example.

All in all it's not my favourite security vs development usability trade-off - specially if you already have server side components (django)

**Comment by edmorley** [comment]
Taskcluster team's decision as to how we use their tokens, not ours fwiw

**Comment by claudijd** [comment]
FWIW, @gdestuynder's point on JS access to LocalStorage and the outcome when we have an XSS is well founded.  Here's a taskcluster example from a few years ago => https://bugzilla.mozilla.org/show_bug.cgi?id=1255147.  Essentially, you're putting a lot of trust that there will never be an XSS issue.  You can increase confidence with CSP, but to me, it still feels unsafe to put secrets in LocalStorage, that said, I'm not that front-end savvy to provide a good alternative ATM.  I will be thinking about this more over the next few days and will report back if I can anything meaningful to add.

**Comment by peterbe** [comment]
I too worry about web storage (local storage and session storage). Seems the best approach is to use cookies with `httpOnly` and make sure you require XSRF protection in the server you're using. mozilla-django-oidc and Django gets this right. Django also have an out-of-the-box session storage solution so you don't even need to put the JWT in the cookie (also the HTTP headers potentially a lot smaller). 

One thing for this project to consider is a sign in flow that does NOT require a callback URL. I.e. the client is self responsible for acquiring a JWT token, but it doesn't want to store it in the client. It'd be better for it to trade it in, with Django, for a session cookie instead. And django can get the expiration of the session cookie to match the expiration of the JWT. Does mozilla-django-oidc support this?

**Comment by peterbe** [comment]
I too worry about web storage (local storage and session storage). Seems the best approach is to use cookies with `httpOnly` and make sure you require XSRF protection in the server you're using. mozilla-django-oidc and Django gets this right. Django also have an out-of-the-box session storage solution so you don't even need to put the JWT in the cookie (also the HTTP headers potentially a lot smaller). 

One thing for this project to consider is a sign in flow that does NOT require a callback URL. I.e. the client is self responsible for acquiring a JWT token, but it doesn't want to store it in the client. It'd be better for it to trade it in, with Django, for a session cookie instead. And django can get the expiration of the session cookie to match the expiration of the JWT. Does mozilla-django-oidc support this?

**Comment by peterbe** [comment]
This issue has gotten out of hand. Let's reign it back in. 

If you use the "code grant" (i.e. the OIDC server redirects to your *server* with a code) then `mozilla-django-oidc` is it and it's wonderful. 
If you want to use the "implicit grant" (i.e. single page app without a Django server to send a code to) then you won't be using `mozilla-django-oidc` and it's technically a non-issue for this project. 

If you do have a Django server to do the initial authentication but then almost all the other stuff you do in the site is via XHR then `mozilla-django-oidc` is quite clunky. The reason is [this](https://github.com/mozilla/mozilla-django-oidc/blob/master/mozilla_django_oidc/middleware.py#L81). The "excuse" is that it's hard to deal with redirects in XHR requests. E.g. 
```javascript
fetch('/server/api/get/stuff', {credentials: 'same-origin'})
.then(response => {
  if (response.status === 200) {
     response.json()
     .then(stuff => {
        doSomethingWith(stuff)
     })
  } else {
     alert("Crap! An error happened!")
  }
})
```
And challenge is that `fetch()` will follow redirects. True for basically any 3xx error. 

Ultimately I think we make a special case for `request.is_ajax()`. If that's the case, and the middleware deems it time to do a session refresh, then it should yield a 403. Why? Because the security of the server wants to say "I don't want to serve you. But if you mind going to <url> to refresh I'd be willing to try again."

Then, the ajax-y sites can watch out for that. E.g.

```javascript
fetch('/server/api/get/stuff', {credentials: 'same-origin'})
.then(response => {
  if (response.status === 403 && response.headers.get("refresh_url")) {
    // This could be made a lot fancier. 
    alert('You have to refresh your authentication session. Going to redirect you.')
    document.location.href = response.headers.get("refresh_url")
  } else if (response.status === 200) {
     response.json()
     .then(stuff => {
        doSomethingWith(stuff)
     })
  } else {
     alert("Crap! An error happened!")
  }
})
```
